Assets/Mirror/Runtime/Transport/LatencySimulation.cs

```diff
@@ -0,0 +1,274 @@
+                    bool drop = random.NextDouble() < unreliableLoss;
+                    if (!drop)
+                    {
+                        // simulate scramble (Random.Next is < max, so +1)
+                        int last = unreliableServerToClient.Count;
+                        int index = unreliableScramble ? random.Next(0, last + 1) : last;
+
+                        // simulate latency
+                        unreliableServerToClient.Insert(index, message);
+                    }
+                    break;
+                default:
+                    Debug.LogError($"{nameof(LatencySimulation)} unexpected channelId: {channelId}");
+                    break;
+            }
+        }
+
+        public override void ServerStart()
+        {
+            wrap.OnServerConnected = OnServerConnected;
+            wrap.OnServerDataReceived = OnServerDataReceived;
+            wrap.OnServerError = OnServerError;
+            wrap.OnServerDisconnected = OnServerDisconnected;
+            wrap.ServerStart();
+        }
+
+        public override void ServerStop()
+        {
+            wrap.ServerStop();
+            reliableServerToClient.Clear();
+            unreliableServerToClient.Clear();
+        }
+
+        public override void ClientEarlyUpdate() => wrap.ClientEarlyUpdate();
+        public override void ServerEarlyUpdate() => wrap.ServerEarlyUpdate();
+        public override void ClientLateUpdate()
+        {
+            // flush reliable messages after latency
+            while (reliableClientToServer.Count > 0)
+            {
+                // check the first message time
+                QueuedMessage message = reliableClientToServer[0];
+                if (message.time + reliableLatency <= Time.time)
+                {
+                    // send and eat
+                    wrap.ClientSend(Channels.DefaultReliable, new ArraySegment<byte>(message.bytes));
+                    reliableClientToServer.RemoveAt(0);
+                }
+                // not enough time elapsed yet
+                break;
+            }
+
+            // flush unreliable messages after latency
+            while (unreliableClientToServer.Count > 0)
+            {
+                // check the first message time
+                QueuedMessage message = unreliableClientToServer[0];
+                if (message.time + unreliableLatency <= Time.time)
+                {
+                    // send and eat (we peeked before)
+                    wrap.ClientSend(Channels.DefaultUnreliable, new ArraySegment<byte>(message.bytes));
+                    unreliableClientToServer.RemoveAt(0);
+                }
+                // not enough time elapsed yet
+                break;
+            }
+
+            // update wrapped transport too
+            wrap.ClientLateUpdate();
+        }
+        public override void ServerLateUpdate()
+        {
+            // flush reliable messages after latency
```

Comment:
Why two loops instead of one that does both reliable and unreliable sends?

Reply:
checking

Reply:
queues need to be separate so we can scramble unreliable messages.
would've been nice though yea